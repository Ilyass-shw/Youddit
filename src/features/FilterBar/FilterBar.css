.filter-bar-container {
	position: relative;
}
.filter-bar {
	justify-content: flex-start;
	align-items: center;
	margin-left: 4.5rem;
	height: 3.5rem;
	border-block-end: 1px solid #404040;
	z-index: 1;
	overflow-y: hidden;
	white-space: nowrap;
	scroll-behavior: smooth;
}
/* Hide scrollbar for Chrome, Safari and Opera */
.filter-bar::-webkit-scrollbar {
	display: none;
}

/* Hide scrollbar for IE, Edge and Firefox */
.filter-bar {
	-ms-overflow-style: none; /* IE and Edge */
	scrollbar-width: none; /* Firefox */
}
.filter-button {
	background-color: #373737;
	color: rgba(255, 255, 255, 1);
	height: 2.2rem;
	border: 1px solid rgba(75, 75, 75, 1);
	margin: 12px;
	margin-left: 0;
	border-radius: 20px;
	padding-right: 8.5px;
	transition: 0.28s;
}

.filter-button:hover {
	background-color: rgba(77, 77, 77, 1);
}
.selected {
	color: #030303;
	background-color: #ffff;
}
.selected:hover {
	background-color: #909090;
}
.arrow {
	position: absolute;
	display: none;
	height: 100%;
	z-index: 2;
}
.hide {
	display: none;
}
.show {
	display: flex;
}
.left {
	left: 0;
}
.left .arrow-button {
	justify-content: flex-start;
	background: linear-gradient(to left, #21212100 -2%, #202020 47%);
}
.right {
	right: 0;
	top: 0;
}
.right .arrow-button {
	justify-content: flex-end;
	background: linear-gradient(to right, #21212100 -2%, #202020 47%);
}
.arrow-button {
	display: flex;
	align-items: center;
	width: 5rem;
	height: 100%;
}
.arrow-button svg {
	padding: 0.5rem;
	border-radius: 50%;
	transition: all 0.28s;
}
.arrow-button svg:hover {
	background-color: #373737;
	cursor: pointer;
}
