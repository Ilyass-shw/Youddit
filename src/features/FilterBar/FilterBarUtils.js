export const isInViewport = (element) => {
	const rect = element.getBoundingClientRect();
	return (
		rect.top >= 0 &&
		rect.left >= 0 &&
		rect.bottom <= (window.innerHeight || document.documentElement.clientHeight) &&
		rect.right <= (window.innerWidth || document.documentElement.clientWidth)
	);
};

export const updateArrowClass = (setShowLeft, setShowRight, firstFilter, lastFilter) => {
	if (!isInViewport(document.getElementById(firstFilter))) {
		setShowLeft(true);
	}
	if (!isInViewport(document.getElementById(lastFilter))) {
		setShowRight(true);
	}
};

export const handleEvent = (setShowLeft, setShowRight, firstFilter, lastFilter) => {
	setShowLeft(false);
	setShowRight(false);
	updateArrowClass(setShowLeft, setShowRight, firstFilter, lastFilter);
};
