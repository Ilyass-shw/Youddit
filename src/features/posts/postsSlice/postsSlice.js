import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import { client, getEndPoint, handlefetchedPosts } from "../../../api/api";

const initialState = {
	searchTerm: "memes",
	postList: [],
	filter: "relevance",
	firstFilter: "relevance",
	lastFilter: "comments",
	postFetchingStatus: "idle",
	error: null,
};

export const fetchPosts = createAsyncThunk("posts/fetchPosts", async (_, { getState }) => {
	const filter = getState().posts.filter;
	const searchTerm = getState().posts.searchTerm;

	const endPoint = getEndPoint(searchTerm, filter);
	const response = await client.get(endPoint);

	const fetchedPosts = handlefetchedPosts(response);
	return fetchedPosts;
});

const postsSlice = createSlice({
	name: "posts",
	initialState,
	reducers: {
		searchTermSet(state, action) {
			state.searchTerm = action.payload;
		},
		filterUpdated(state, action) {
			state.filter = action.payload;
		},
		firstFilterUpdated(state, action) {
			state.firstFilter = action.payload;
		},
		lastFilterUpdated(state, action) {
			state.lastFilter = action.payload;
		},
	},
	extraReducers: {
		[fetchPosts.pending]: (state, action) => {
			state.postFetchingStatus = "Loading";
		},
		[fetchPosts.fulfilled]: (state, action) => {
			state.postFetchingStatus = "succeeded";
			state.postList = action.payload;
		},
		[fetchPosts.rejected]: (state, action) => {
			state.postFetchingStatus = "failed";
			state.error = action.error.message;
		},
	},
});

export const { searchTermSet, filterUpdated, firstFilterUpdated, lastFilterUpdated } = postsSlice.actions;
export default postsSlice.reducer;

export const selectAllPosts = (state) => state.posts.postList;
