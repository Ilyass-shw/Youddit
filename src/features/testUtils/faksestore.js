import { configureStore } from "@reduxjs/toolkit";
import { createSlice } from "@reduxjs/toolkit";

const initialState = {
	firstFilter: "relevance",
	lastFilter: "comments",
	filter: "relevance",
};
const fakeSlice = createSlice({
	name: "posts",
	initialState,
	reducers: {},
});

export default configureStore({
	reducer: {
		posts: fakeSlice.reducer,
	},
});

export const makeCustomTestStore = (name, initialState, reducers = {}) => {
	const fakeSlice = createSlice({
		name,
		initialState,
		reducers,
	});

	return configureStore({
		reducer: {
			posts: fakeSlice.reducer,
		},
	});
};
