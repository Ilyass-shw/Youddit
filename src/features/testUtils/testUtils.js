import React from "react";
import { render as rtlRender } from "@testing-library/react";
import { Provider } from "react-redux";
import store, { makeCustomTestStore } from "./faksestore";

const TestProvider = ({ store, children }) => <Provider store={store}>{children}</Provider>;

export function render(ui, { store, ...otherOpts }) {
	return rtlRender(<TestProvider store={store}>{ui}</TestProvider>, otherOpts);
}

export function makeTestStore(name, initialState, reducers) {
	if (name && initialState) {
		const Store = makeCustomTestStore(name, initialState, reducers);
		const origDispatch = Store.dispatch;
		Store.dispatch = jest.fn(origDispatch);
		return Store;
	} else {
		const origDispatch = store.dispatch;
		store.dispatch = jest.fn(origDispatch);
		return store;
	}
}

// re-export everything
export * from "@testing-library/react";
