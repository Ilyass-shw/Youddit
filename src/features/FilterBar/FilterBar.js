import React, { useEffect, useState, useRef } from "react";
import { useSelector } from "react-redux";
import "./FilterBar.css";
import { updateArrowClass, handleEvent } from "./FilterBarUtils";
import FilterButtons from "../FilterButtons/FilterButtons";

import { InlineIcon } from "@iconify/react";

import arrowIosForwardOutline from "@iconify-icons/eva/arrow-ios-forward-outline";
import arrowIosBackOutline from "@iconify-icons/eva/arrow-ios-back-outline";

export const FilterBar = () => {
	const [showLeft, setShowLeft] = useState(false);
	const [showRight, setShowRight] = useState(false);

	const firstFilter = useSelector((state) => state.posts.firstFilter);
	const lastFilter = useSelector((state) => state.posts.lastFilter);

	const filterBar = useRef();

	const handleScrollLeft = () => {
		filterBar.current.scrollLeft -= 100;
	};

	const handleScrollRight = () => {
		filterBar.current.scrollLeft += 100;
	};
	// style scroll arrows on first render
	useEffect(() => {
		updateArrowClass(setShowLeft, setShowRight, firstFilter, lastFilter);
	}, [firstFilter, lastFilter]);

	// style scroll arrows onScroll and onResize
	useEffect(() => {
		const FilterBar = filterBar.current;
		FilterBar.addEventListener("scroll", () => {
			handleEvent(setShowLeft, setShowRight, firstFilter, lastFilter);
		});
		window.addEventListener("resize", () => {
			handleEvent(setShowLeft, setShowRight, firstFilter, lastFilter);
		});

		return () => {
			FilterBar.removeEventListener("scroll", () => {
				handleEvent(setShowLeft, setShowRight, firstFilter, lastFilter);
			});
			window.addEventListener("resize", () => {
				handleEvent(setShowLeft, setShowRight, firstFilter, lastFilter);
			});
		};
	}, [firstFilter, lastFilter]);

	return (
		<div className="filter-bar-container">
			<div ref={filterBar} className="filter-bar">
				<div className={`arrow left ${showLeft ? "show" : "hide"}`}>
					<div className="arrow-button" data-testid="left-arrow" onClick={handleScrollLeft}>
						<InlineIcon icon={arrowIosBackOutline} />
					</div>
				</div>

				<FilterButtons />

				<div className={`arrow right  ${showRight ? "show" : "hide"}`}>
					<div className="arrow-button" data-testid="right-arrow" onClick={handleScrollRight}>
						<InlineIcon icon={arrowIosForwardOutline} />
					</div>
				</div>
			</div>
		</div>
	);
};
