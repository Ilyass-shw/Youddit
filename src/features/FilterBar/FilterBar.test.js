import React from "react";
import { FilterBar } from "./FilterBar";
import { jest } from "@jest/globals";
import { filterUpdated } from "../posts/postsSlice/postsSlice";
import { render, makeTestStore, fireEvent, screen } from "../testUtils/testUtils";
jest.mock("@iconify/react", () => {
	return { InlineIcon: () => <p>InlineIcon mock</p> };
});

jest.mock("@iconify-icons/bx/bxs-rocket", () => {
	return {
		default: () => {
			return "rock";
		},
	};
});
jest.mock("@iconify-icons/ant-design/fire-filled", () => {
	return {
		default: () => {
			return "rock";
		},
	};
});
jest.mock("@iconify-icons/ic/sharp-new-releases", () => {
	return {
		default: () => {
			return "rock";
		},
	};
});
jest.mock("@iconify/icons-oi/vertical-align-top", () => {
	return {
		default: () => {
			return "rock";
		},
	};
});
jest.mock("@iconify/icons-icons8/comments", () => {
	return {
		default: () => {
			return "rock";
		},
	};
});
jest.mock("@iconify-icons/eva/arrow-ios-forward-outline", () => {
	return {
		default: () => {
			return "rock";
		},
	};
});
jest.mock("@iconify-icons/eva/arrow-ios-back-outline", () => {
	return {
		default: () => {
			return "rock";
		},
	};
});

let store;

describe("FilterBar", () => {
	beforeEach(() => {
		store = makeTestStore();
	});

	it("should match the snapshot", () => {
		const { container } = render(<FilterBar />, { store });
		expect(container).toMatchSnapshot();
	});

	it("should render without crashing", () => {
		render(<FilterBar />, { store });
	});

	it("should render with 'Relevance' as the default selected filter", () => {
		render(<FilterBar />, { store });

		const relevanceFilterButton = screen.getByText("Relevance");

		expect(relevanceFilterButton).toHaveClass("filter-button selected");
	});

	it("should be able to select any button as the selected filter and one at a time", () => {
		render(<FilterBar />, { store });

		const relevanceFilterButton = screen.getByText("Relevance");
		const hotFilterButton = screen.getByText("Hot");
		const newFilterButton = screen.getByText("New");
		const commentsFilterButton = screen.getByText("Comments");
		const topFilterButton = screen.getByText("Top");

		expect(hotFilterButton).toHaveClass("filter-button false");
		expect(newFilterButton).toHaveClass("filter-button false");
		expect(commentsFilterButton).toHaveClass("filter-button false");
		expect(topFilterButton).toHaveClass("filter-button false");
		expect(relevanceFilterButton).toHaveClass("filter-button selected");

		fireEvent.click(hotFilterButton);

		expect(hotFilterButton).toHaveClass("filter-button selected");
		expect(newFilterButton).toHaveClass("filter-button false");
		expect(commentsFilterButton).toHaveClass("filter-button false");
		expect(topFilterButton).toHaveClass("filter-button false");
		expect(relevanceFilterButton).toHaveClass("filter-button false");

		fireEvent.click(newFilterButton);

		expect(hotFilterButton).toHaveClass("filter-button false");
		expect(newFilterButton).toHaveClass("filter-button selected");
		expect(commentsFilterButton).toHaveClass("filter-button false");
		expect(topFilterButton).toHaveClass("filter-button false");
		expect(relevanceFilterButton).toHaveClass("filter-button false");

		fireEvent.click(commentsFilterButton);

		expect(hotFilterButton).toHaveClass("filter-button false");
		expect(newFilterButton).toHaveClass("filter-button false");
		expect(commentsFilterButton).toHaveClass("filter-button selected");
		expect(topFilterButton).toHaveClass("filter-button false");
		expect(relevanceFilterButton).toHaveClass("filter-button false");

		fireEvent.click(topFilterButton);

		expect(hotFilterButton).toHaveClass("filter-button false");
		expect(newFilterButton).toHaveClass("filter-button false");
		expect(commentsFilterButton).toHaveClass("filter-button false");
		expect(topFilterButton).toHaveClass("filter-button selected");
		expect(relevanceFilterButton).toHaveClass("filter-button false");

		fireEvent.click(relevanceFilterButton);

		expect(hotFilterButton).toHaveClass("filter-button false");
		expect(newFilterButton).toHaveClass("filter-button false");
		expect(commentsFilterButton).toHaveClass("filter-button false");
		expect(topFilterButton).toHaveClass("filter-button false");
		expect(relevanceFilterButton).toHaveClass("filter-button selected");
	});

	it("should render with the relative filter everytime a new filter button is clicked", () => {
		render(<FilterBar />, { store });

		const hotFilterButton = screen.getByText("Hot");
		fireEvent.click(hotFilterButton);

		expect(store.dispatch).toHaveBeenCalledWith(filterUpdated("hot"));

		const TopFilterButton = screen.getByText("Top");
		fireEvent.click(TopFilterButton);

		expect(store.dispatch).toHaveBeenCalledWith(filterUpdated("top"));

		const NewFilterButton = screen.getByText("New");
		fireEvent.click(NewFilterButton);

		expect(store.dispatch).toHaveBeenCalledWith(filterUpdated("new"));

		const RelevanceFilterButton = screen.getByText("Relevance");
		fireEvent.click(RelevanceFilterButton);

		expect(store.dispatch).toHaveBeenCalledWith(filterUpdated("relevance"));

		const CommentsFilterButton = screen.getByText("Comments");
		fireEvent.click(CommentsFilterButton);

		expect(store.dispatch).toHaveBeenCalledWith(filterUpdated("comments"));
	});
});
