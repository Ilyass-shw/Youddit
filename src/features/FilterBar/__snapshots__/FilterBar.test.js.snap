// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterBar should match the snapshot 1`] = `
<div>
  <div
    class="filter-bar-container"
  >
    <div
      class="filter-bar"
    >
      <div
        class="arrow left hide"
      >
        <div
          class="arrow-button"
          data-testid="left-arrow"
        >
          <p>
            InlineIcon mock
          </p>
        </div>
      </div>
      <div>
        <button
          class="filter-button selected"
          id="relevance"
          value="relevance"
        >
          <p>
            InlineIcon mock
          </p>
           Relevance
        </button>
        <button
          class="filter-button false"
          id="hot"
          value="hot"
        >
          <p>
            InlineIcon mock
          </p>
           Hot
        </button>
        <button
          class="filter-button false"
          id="new"
          value="new"
        >
          <p>
            InlineIcon mock
          </p>
           New
        </button>
        <button
          class="filter-button false"
          id="top"
          value="top"
        >
          <p>
            InlineIcon mock
          </p>
           Top
        </button>
        <button
          class="filter-button false"
          id="comments"
          value="comments"
        >
          <p>
            InlineIcon mock
          </p>
           Comments
        </button>
      </div>
      <div
        class="arrow right  hide"
      >
        <div
          class="arrow-button"
          data-testid="right-arrow"
        >
          <p>
            InlineIcon mock
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
